### PR TITLE
tree-wide: wipe alloca() from the codebase

### PR DESCRIPTION
**Pull request by brauner**
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

File: src/include/lxcmntent.c [hunk 1]
```diff
@@ -154,20 +154,24 @@ struct mntent *getmntent(FILE *stream)
...
 {
 	/* Extend the mode parameter with "c" to disable cancellation in the
 	 * I/O functions and "e" to set FD_CLOEXEC.
 	 */
 	size_t modelen = strlen(mode);
-	char *newmode;
+	char newmode[__SETMNTENT_MODE_MAX];
 
-	newmode = alloca(modelen + 3);
+	if (modelen >= (__SETMNTENT_MODE_MAX - 2)) {
```
Comment (Blub) [inline src/include/lxcmntent.c]: The `strlen()` doesn't include the `\0` so this should be `3` - equivalent to the `"ce\0"` we append.

File: src/lxc/cgroups/cgfsng.c [hunk 2]
```diff
@@ -2209,15 +2213,12 @@ __cgfsng_ops static int cgfsng_get(struct cgroup_ops *ops, const char *filename,
 				     char *value, size_t len, const char *name,
 				     const char *lxcpath)
 {
-	int ret = -1;
-	size_t controller_len;
-	char *controller, *p, *path;
+	__attribute__((__cleanup__(__auto_free__))) char *controller;
```
Comment (Blub) [inline src/lxc/cgroups/cgfsng.c]: Should we add a macro to shorten this to memory_utils.h? `lxc_autofree`?

File: src/include/lxcmntent.c (hunk 1 above)
Comment (brauner) [reply src/include/lxcmntent.c]: Thanks for spotting. Can you go through all of it now, please? 😬

File: src/lxc/cmd/lxc_user_nic.c [hunk 3]
```diff
@@ -838,13 +840,10 @@ static char *get_nic_if_avail(int fd, struct alloted_s *names, int pid,
...
 {
+	__attribute__((__cleanup__(__auto_free__))) char *p;
 	int ret;
-	char *p;
-	size_t len;
 
-	len = strlen(fnam);
-	p = alloca(len + 1);
-	(void)strlcpy(p, fnam, len + 1);
+	must_copy_string(fnam);
```
Comment (Blub) [inline src/lxc/cmd/lxc_user_nic.c]: Missing `p =` ;-)

Comment (brauner) [reply src/lxc/cmd/lxc_user_nic.c]: Good catch!

**Comment by brauner** [comment]
Note that I wrote this when I got home from Brussels and with zero sleep. So @Blub be extra cautious. 😬

File: src/lxc/start.c [hunk 4]
```diff
@@ -97,14 +97,16 @@ static void lxc_destroy_container_on_signal(struct lxc_handler *handler,
...
+	__attribute__((__cleanup__(__auto_free__))) char *copy;
 	int ret;
 	size_t len;
-	char *copy, *e, *p, saved;
+	char *e, *p, saved;
 
 	len = strlen(path);
-	copy = alloca(len + 1);
+	copy = must_realloc(NULL, len + 1);
 	(void)strlcpy(copy, path, len + 1);
```
Comment (Blub) [inline src/lxc/start.c]: The above 2 lines should be removed with the `must_copy_string()` around below.

Comment (brauner) [reply src/lxc/start.c]: done

File: src/lxc/cgroups/cgfsng.c (hunk 2 above)
Comment (brauner) [reply src/lxc/cgroups/cgfsng.c]: done

**Comment by brauner** [comment]
That's why it's marked as `incomplete`. :)

**Comment by brauner** [comment]
@Blub, @hallyn, someone please merge. :)

File: src/lxc/namespace.c [hunk 5]
```diff
@@ -53,16 +53,17 @@ static int do_clone(void *arg)
...
+#define __LXC_STACK_SIZE 4096
 pid_t lxc_clone(int (*fn)(void *), void *arg, int flags)
 {
+	size_t stack_size;
+	pid_t ret;
 	struct clone_arg clone_arg = {
-		.fn = fn,
-		.arg = arg,
+	    .fn = fn,
+	    .arg = arg,
```
Comment (Blub) [inline src/lxc/namespace.c]: Why the space change?

File: src/lxc/namespace.c [hunk 6]
```diff
@@ -53,16 +53,17 @@ static int do_clone(void *arg)
...
 	};
-
-	size_t stack_size = lxc_getpagesize();
-	void *stack = alloca(stack_size);
-	pid_t ret;
+	char *stack[__LXC_STACK_SIZE] = {0};
```
Comment (Blub) [inline src/lxc/namespace.c]: The commit message should probably explain that (and why) this also changes from `getpagesize()` to a hardcoded `4096`.

File: src/lxc/namespace.c (hunk 5 above)
Comment (brauner) [reply src/lxc/namespace.c]: clang-format's the culprit.

Comment (brauner) [reply src/lxc/namespace.c]: but all variables have been reordered anyway so let's just leave it.

File: src/lxc/namespace.c (hunk 6 above)
Comment (brauner) [reply src/lxc/namespace.c]: Avoiding allocations
